Source/Core/Core/PowerPC/Jit64/Jit64_Util.cpp

```diff
@@ -0,0 +1,420 @@
+
+void Jit64::SafeWrite(GPRRegister& reg_value, GPRRegister& reg_addr, GPRRegister& offset,
+                      int accessSize, bool swap, bool update)
+{
+  _assert_msg_(DYNAREC, accessSize == 8 || accessSize == 16 || accessSize == 32 || accessSize == 64,
+               "Invalid write size: %d", accessSize);
+
+  if (update && !reg_addr.HasPPCRegister())
+  {
+    // This form of instruction is illegal
+    WriteInvalidInstruction();
+    return;
+  }
+
+  bool valueAddressShareRegister = reg_value.IsAliasOf(reg_addr);
+  bool valueOffsetShareRegister = reg_value.IsAliasOf(offset);
+
+  // We want to preserve any non-scratch registers passed into this
+  // function. If a scratch register is passed, we assume it may be safely
+  // clobbered.
+  if (WriteClobbersRegValue(accessSize, swap) && reg_value.HasPPCRegister())
+  {
+    // TODO: test for free register, otherwise swap back
+    auto scratch = reg_value.BorrowCopy();
+    SafeWrite(scratch, reg_addr, offset, accessSize, swap, update);
+    return;
+  }
+
+  // Constant writes have a much faster path
+  if (reg_addr.IsImm() && offset.IsImm())
+  {
+    u32 const_addr = reg_addr.Imm32() + offset.Imm32();
+    if (UnsafeWriteToConstAddress(accessSize, reg_value, const_addr, swap))
+    {
+      if (update)
+        reg_addr.SetImm32(const_addr);
+      return;
+    }
+
+    // Nope, we'll fall back to the standard write...
+  }
+
+  // Note that while SafeWriteRegToReg takes a register and an OpArg, it really wants things to be
+  // registers or immediates.
+  auto val =
+      (reg_value.IsImm() || reg_value.IsRegBound()) ? reg_value : reg_value.Bind(BindMode::Read);
+
+  // Optimize swap flag
+  if (swap)
+  {
+    // No need to swap
+    if (accessSize == 8)
+    {
+      swap = false;
+    }
+    else
+    {
+      // We can swap these ahead of time
+      if (val.IsImm())
+      {
+        swap = false;
+        switch (accessSize)
+        {
+        case 16:
+          val = regs.gpr.Imm32(Common::swap16(val.Imm32()));
+          break;
+        case 32:
+          val = regs.gpr.Imm32(Common::swap32(val.Imm32()));
+          break;
+        case 64:
+          // We don't support 64-bit immediates (TODO?)
+          UnexpectedInstructionForm();
+          break;
+        }
+      }
+    }
+  }
+  u32 flags = swap ? 0 : SAFE_LOADSTORE_NO_SWAP;
+
+  // We need to bind a register, so may as well just add the offset here
+  if (reg_addr.IsImm())
+  {
+    if (update)
+    {
+      // Update reg_addr inside a transaction
+      auto xa = reg_addr.Bind(BindMode::ReadWriteTransactional);
+      MOV_sum(32, xa, xa, offset);
+      SafeWriteRegToReg(val, xa, accessSize, 0, CallerSavedRegistersInUse(), flags);
+    }
+    else
+    {
+      auto scratch = regs.gpr.Borrow();
+      MOV_sum(32, scratch, reg_addr, offset);
+      SafeWriteRegToReg(val, scratch, accessSize, 0, CallerSavedRegistersInUse(), flags);
+    }
+    return;
```

Comment:
Is this point of special-casing immediate reg_addr just to avoid materializing the immediate (`auto addr = reg_addr.Bind(BindMode::Read);`)?  It seems like you could refactor this to avoid code duplication... or at least add a comment explicitly mentioning it.